File: udon_decompiler/analysis/transform/passes/temp_variable_inline.py
from __future__ import annotations

from copy import deepcopy

from udon_decompiler.analysis.ir.nodes import (
    IRAssignmentStatement,
    IRBlock,
    IRBlockContainer,
    IRConstructorCallExpression,
    IRExpression,
    IRExpressionStatement,
    IRExternalCallExpression,
    IRHighLevelDoWhile,
    IRHighLevelSwitch,
    IRHighLevelWhile,
    IRIf,
    IRJump,
    IRLeave,
    IROperatorCallExpression,
    IRPropertyAccessExpression,
    IRReturn,
    IRStatement,
    IRSwitch,
    IRVariableExpression,
)
from udon_decompiler.analysis.transform.ir_utils import (
    get_block_terminator,
    iter_block_containers,
    iter_block_targets,
)
from udon_decompiler.analysis.transform.pass_base import (
    IILTransform,
    ILTransformContext,
)
from udon_decompiler.models.program import SymbolInfo


class TempVariableInline(IILTransform):
    """Inline adjacent single-use ``__intnl_*`` temporaries within one block."""

    def run(self, function, context: ILTransformContext) -> None:
        self._function = function
        self._successor_cache: dict[int, dict[IRBlock, list[IRBlock]]] = {}
        visited_blocks: set[int] = set()

        for container in iter_block_containers(function):
            for block in container.blocks:
                self._inline_block(block, container, visited_blocks)

    def _inline_block(
        self,
        block: IRBlock,
        container: IRBlockContainer,
        visited_blocks: set[int],
    ) -> None:
        block_id = id(block)
        if block_id in visited_blocks:
            return
        visited_blocks.add(block_id)

        i = 1
        while i < len(block.statements):
            if self._try_inline_adjacent(block, container, i):
                i = max(1, i - 1)
                continue
            i += 1

        for statement in block.statements:
            self._inline_nested(statement, container, visited_blocks)

    def _inline_nested(
        self,
        statement: IRStatement,
        container: IRBlockContainer,
        visited_blocks: set[int],
    ) -> None:
        if isinstance(statement, IRBlock):
            self._inline_block(statement, container, visited_blocks)
            return

        if isinstance(statement, IRBlockContainer):
            for block in statement.blocks:
                self._inline_block(block, statement, visited_blocks)
            return

        if isinstance(statement, IRIf):
            self._inline_nested(
                statement.true_statement,
                container,
                visited_blocks,
            )
            if statement.false_statement is not None:
                self._inline_nested(
                    statement.false_statement,
                    container,
                    visited_blocks,
                )

    def _try_inline_adjacent(
        self,
        block: IRBlock,
        container: IRBlockContainer,
        index: int,
    ) -> bool:
        prev = block.statements[index - 1]
        curr = block.statements[index]

        if not isinstance(prev, IRAssignmentStatement):
            return False

        variable = prev.target
        if not variable.name.startswith(SymbolInfo.INTERNAL_SYMBOL_PREFIX):
            return False

        if self._writes_variable_top_level(curr, variable):
            return False

        if self._count_reads_top_level(curr, variable) != 1:
            return False

        # Keep the defining assignment when its value can flow to a read in
        # another block before any redefinition.
        if self._definition_flows_to_other_block_read(
            container=container,
            block=block,
            index=index,
            variable=variable,
        ):
            return False

        if not self._replace_reads_top_level(curr, variable, prev.value):
            return False

        block.statements.pop(index - 1)
        return True

    def _definition_flows_to_other_block_read(
        self,
        container: IRBlockContainer,
        block: IRBlock,
        index: int,
        variable,
    ) -> bool:
        # First scan tail in current block.
        for statement in block.statements[index + 1 :]:
            if self._count_reads_in_statement(statement, variable) > 0:
                return True
            if self._count_writes_in_statement(statement, variable) > 0:
                return False

        # Then follow CFG edges to detect reads before next write.
        successors = self._get_successor_map(container)
        worklist = list(successors.get(block, []))
        visited: set[IRBlock] = set()

        while worklist:
            current = worklist.pop()
            if current in visited:
                continue
            visited.add(current)

            killed = False
            for statement in current.statements:
                if self._count_reads_in_statement(statement, variable) > 0:
                    return True
                if self._count_writes_in_statement(statement, variable) > 0:
                    killed = True
                    break

            if not killed:
                worklist.extend(successors.get(current, []))

        return False

    def _get_successor_map(
        self,
        container: IRBlockContainer,
    ) -> dict[IRBlock, list[IRBlock]]:
        key = id(container)
        cached = self._successor_cache.get(key)
        if cached is not None:
            return cached

        successors: dict[IRBlock, list[IRBlock]] = {}
        block_set = set(container.blocks)

        for index, block in enumerate(container.blocks):
            succ_list: list[IRBlock] = []
            next_block = (
                container.blocks[index + 1]
                if index + 1 < len(container.blocks)
                else None
            )
            terminator = get_block_terminator(block)

            if terminator is None:
                if next_block is not None:
                    succ_list.append(next_block)
                successors[block] = succ_list
                continue

            for target in iter_block_targets(terminator):
                if target in block_set and target not in succ_list:
                    succ_list.append(target)

            if isinstance(terminator, IRIf) and terminator.false_statement is None:
                if next_block is not None and next_block not in succ_list:
                    succ_list.append(next_block)
            elif isinstance(terminator, IRSwitch) and terminator.default_target is None:
                if next_block is not None and next_block not in succ_list:
                    succ_list.append(next_block)

            successors[block] = succ_list

        self._successor_cache[key] = successors
        return successors

    def _count_reads_in_statement(self, statement: IRStatement, variable) -> int:
        if isinstance(statement, IRAssignmentStatement):
            return self._count_expr_reads(statement.value, variable)

        if isinstance(statement, IRExpressionStatement):
            return self._count_expr_reads(statement.expression, variable)

        if isinstance(statement, IRIf):
            count = self._count_expr_reads(statement.condition, variable)
            count += self._count_reads_in_statement(statement.true_statement, variable)
            if statement.false_statement is not None:
                count += self._count_reads_in_statement(
                    statement.false_statement,
                    variable,
                )
            return count

        if isinstance(statement, IRSwitch):
            return self._count_expr_reads(statement.index_expression, variable)

        if isinstance(statement, IRBlock):
            count = 0
            for nested in statement.statements:
                count += self._count_reads_in_statement(nested, variable)
            return count

        if isinstance(statement, IRBlockContainer):
            count = 0
            for nested_block in statement.blocks:
                count += self._count_reads_in_statement(nested_block, variable)
            return count

        if isinstance(statement, IRHighLevelSwitch):
            count = self._count_expr_reads(statement.index_expression, variable)
            for section in statement.sections:
                count += self._count_reads_in_statement(section.body, variable)
            return count

        if isinstance(statement, (IRHighLevelWhile, IRHighLevelDoWhile)):
            return self._count_reads_in_statement(statement.body, variable)

        if isinstance(statement, (IRJump, IRLeave, IRReturn)):
            return 0

        return 0

    def _count_writes_in_statement(self, statement: IRStatement, variable) -> int:
        if isinstance(statement, IRAssignmentStatement):
            return 1 if statement.target == variable else 0

        if isinstance(statement, IRIf):
            count = self._count_writes_in_statement(statement.true_statement, variable)
            if statement.false_statement is not None:
                count += self._count_writes_in_statement(
                    statement.false_statement,
                    variable,
                )
            return count

        if isinstance(statement, IRBlock):
            count = 0
            for nested in statement.statements:
                count += self._count_writes_in_statement(nested, variable)
            return count

        if isinstance(statement, IRBlockContainer):
            count = 0
            for nested_block in statement.blocks:
                count += self._count_writes_in_statement(nested_block, variable)
            return count

        if isinstance(statement, IRHighLevelSwitch):
            count = 0
            for section in statement.sections:
                count += self._count_writes_in_statement(section.body, variable)
            return count

        if isinstance(statement, (IRHighLevelWhile, IRHighLevelDoWhile)):
            return self._count_writes_in_statement(statement.body, variable)

        return 0

    def _writes_variable_top_level(self, statement: IRStatement, variable) -> bool:
        return (
            isinstance(statement, IRAssignmentStatement)
            and statement.target == variable
        )

    def _count_reads_top_level(self, statement: IRStatement, variable) -> int:
        if isinstance(statement, IRAssignmentStatement):
            return self._count_expr_reads(statement.value, variable)

        if isinstance(statement, IRExpressionStatement):
            return self._count_expr_reads(statement.expression, variable)

        if isinstance(statement, IRIf):
            return self._count_expr_reads(statement.condition, variable)

        if isinstance(statement, IRSwitch):
            return self._count_expr_reads(statement.index_expression, variable)

        return 0

    def _count_expr_reads(self, expression: IRExpression, variable) -> int:
        if isinstance(expression, IRVariableExpression):
            return 1 if expression.variable == variable else 0

        if isinstance(expression, IROperatorCallExpression):
            return sum(
                self._count_expr_reads(arg, variable) for arg in expression.arguments
            )

        if isinstance(expression, IRExternalCallExpression):
            return sum(
                self._count_expr_reads(arg, variable) for arg in expression.arguments
            )

        if isinstance(expression, IRPropertyAccessExpression):
            return sum(
                self._count_expr_reads(arg, variable) for arg in expression.arguments
            )

        if isinstance(expression, IRConstructorCallExpression):
            return sum(
                self._count_expr_reads(arg, variable) for arg in expression.arguments
            )

        return 0

    def _replace_reads_top_level(
        self,
        statement: IRStatement,
        variable,
        replacement: IRExpression,
    ) -> bool:
        if isinstance(statement, IRAssignmentStatement):
            statement.value, count = self._replace_expr_reads(
                statement.value,
                variable,
                replacement,
            )
            return count == 1

        if isinstance(statement, IRExpressionStatement):
            statement.expression, count = self._replace_expr_reads(
                statement.expression,
                variable,
                replacement,
            )
            return count == 1

        if isinstance(statement, IRIf):
            statement.condition, count = self._replace_expr_reads(
                statement.condition,
                variable,
                replacement,
            )
            return count == 1

        if isinstance(statement, IRSwitch):
            statement.index_expression, count = self._replace_expr_reads(
                statement.index_expression,
                variable,
                replacement,
            )
            return count == 1

        return False

    def _replace_expr_reads(
        self,
        expression: IRExpression,
        variable,
        replacement: IRExpression,
    ) -> tuple[IRExpression, int]:
        if isinstance(expression, IRVariableExpression):
            if expression.variable == variable:
                return deepcopy(replacement), 1
            return expression, 0

        if isinstance(expression, IROperatorCallExpression):
            count = 0
            new_args: list[IRExpression] = []
            for arg in expression.arguments:
                new_arg, c = self._replace_expr_reads(arg, variable, replacement)
                new_args.append(new_arg)
                count += c
            expression.arguments = new_args
            return expression, count

        if isinstance(expression, IRExternalCallExpression):
            count = 0
            new_args: list[IRExpression] = []
            for arg in expression.arguments:
                new_arg, c = self._replace_expr_reads(arg, variable, replacement)
                new_args.append(new_arg)
                count += c
            expression.arguments = new_args
            return expression, count

        if isinstance(expression, IRPropertyAccessExpression):
            count = 0
            new_args: list[IRExpression] = []
            for arg in expression.arguments:
                new_arg, c = self._replace_expr_reads(arg, variable, replacement)
                new_args.append(new_arg)
                count += c
            expression.arguments = new_args
            return expression, count

        if isinstance(expression, IRConstructorCallExpression):
            count = 0
            new_args: list[IRExpression] = []
            for arg in expression.arguments:
                new_arg, c = self._replace_expr_reads(arg, variable, replacement)
                new_args.append(new_arg)
                count += c
            expression.arguments = new_args
            return expression, count

        return expression, 0

File: src/decompiler/transform/passes/temp_variable_inline.rs
use std::collections::{HashMap, HashSet, VecDeque};

use crate::decompiler::ir::{
    IrAssignmentStatement, IrBlock, IrBlockContainer, IrConstructorCallExpression, IrExpression,
    IrExpressionStatement, IrExternalCallExpression, IrFunction, IrHighLevelDoWhile,
    IrHighLevelSwitch, IrHighLevelWhile, IrIf, IrOperatorCallExpression,
    IrPropertyAccessExpression, IrReturn, IrStatement, IrSwitch,
};
use crate::decompiler::transform::ir_utils::{get_block_terminator, iter_block_targets};
use crate::decompiler::transform::pass_base::{ITransform, TransformContext};
use crate::decompiler::{Result, VariableRecord};

pub struct TempVariableInline;

impl ITransform for TempVariableInline {
    fn run(&self, function: &mut IrFunction, context: &mut TransformContext<'_, '_>) -> Result<()> {
        let mut successor_cache = HashMap::<u32, HashMap<u32, Vec<u32>>>::new();
        let variables = context
            .program_context
            .decompile_context
            .variables
            .variables
            .iter()
            .map(|x| (x.address, x))
            .collect::<HashMap<_, _>>();
        inline_container(&mut function.body, &mut successor_cache, &variables);
        Ok(())
    }
}

fn inline_container(
    container: &mut IrBlockContainer,
    successor_cache: &mut HashMap<u32, HashMap<u32, Vec<u32>>>,
    variables: &HashMap<u32, &VariableRecord>,
) {
    let mut index = 0usize;
    while index < container.blocks.len() {
        let snapshot = container.clone();
        let Some(block) = container.blocks.get_mut(index) else {
            break;
        };
        inline_block(block, &snapshot, successor_cache, variables);
        index += 1;
    }
}

fn inline_block(
    block: &mut IrBlock,
    container: &IrBlockContainer,
    successor_cache: &mut HashMap<u32, HashMap<u32, Vec<u32>>>,
    variables: &HashMap<u32, &VariableRecord>,
) {
    let mut index = 1usize;
    while index < block.statements.len() {
        if try_inline_adjacent(block, container, index, successor_cache, variables) {
            index = index.saturating_sub(1).max(1);
            continue;
        }
        index += 1;
    }

    for statement in &mut block.statements {
        inline_nested(statement, container, successor_cache, variables);
    }
}

fn inline_nested(
    statement: &mut IrStatement,
    container: &IrBlockContainer,
    successor_cache: &mut HashMap<u32, HashMap<u32, Vec<u32>>>,
    variables: &HashMap<u32, &VariableRecord>,
) {
    match statement {
        IrStatement::Block(block) => inline_block(block, container, successor_cache, variables),
        IrStatement::BlockContainer(nested) => inline_container(nested, successor_cache, variables),
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            inline_nested(true_statement, container, successor_cache, variables);
            if let Some(false_statement) = false_statement.as_mut() {
                inline_nested(false_statement, container, successor_cache, variables);
            }
        }
        _ => {}
    }
}

fn try_inline_adjacent(
    block: &mut IrBlock,
    container: &IrBlockContainer,
    index: usize,
    successor_cache: &mut HashMap<u32, HashMap<u32, Vec<u32>>>,
    variables: &HashMap<u32, &VariableRecord>,
) -> bool {
    let (target_address, replacement) = {
        let Some(prev_stmt) = block.statements.get(index - 1) else {
            return false;
        };
        let Some(curr_stmt) = block.statements.get(index) else {
            return false;
        };

        let IrStatement::Assignment(IrAssignmentStatement { target, value }) = prev_stmt else {
            return false;
        };
        let Some(target_address) = assignment_target_variable_address(target) else {
            return false;
        };
        let Some(target) = variables.get(&target_address) else {
            return false;
        };
        if !target.name.starts_with("__intnl_") {
            return false;
        }
        if writes_variable_top_level(curr_stmt, target_address) {
            return false;
        }
        if count_reads_top_level(curr_stmt, target_address) != 1 {
            return false;
        }
        if definition_flows_to_other_block_read(
            container,
            block,
            index,
            target_address,
            successor_cache,
        ) {
            return false;
        }

        (target_address, value.clone())
    };

    let Some(curr_stmt_mut) = block.statements.get_mut(index) else {
        return false;
    };
    if !replace_reads_top_level(curr_stmt_mut, target_address, &replacement) {
        return false;
    }
    block.statements.remove(index - 1);
    true
}

fn definition_flows_to_other_block_read(
    container: &IrBlockContainer,
    block: &IrBlock,
    index: usize,
    variable_address: u32,
    successor_cache: &mut HashMap<u32, HashMap<u32, Vec<u32>>>,
) -> bool {
    for statement in block.statements.iter().skip(index + 1) {
        if count_reads_in_statement(statement, variable_address) > 0 {
            return true;
        }
        if count_writes_in_statement(statement, variable_address) > 0 {
            return false;
        }
    }

    let successors = get_successor_map(container, successor_cache);
    let mut worklist = VecDeque::<u32>::new();
    if let Some(nexts) = successors.get(&block.start_address) {
        for addr in nexts {
            worklist.push_back(*addr);
        }
    }

    let mut visited = HashSet::<u32>::new();
    while let Some(current_address) = worklist.pop_front() {
        if !visited.insert(current_address) {
            continue;
        }

        let Some(current_block) = container
            .blocks
            .iter()
            .find(|x| x.start_address == current_address)
        else {
            continue;
        };

        let mut killed = false;
        for statement in &current_block.statements {
            if count_reads_in_statement(statement, variable_address) > 0 {
                return true;
            }
            if count_writes_in_statement(statement, variable_address) > 0 {
                killed = true;
                break;
            }
        }
        if !killed && let Some(nexts) = successors.get(&current_address) {
            for addr in nexts {
                worklist.push_back(*addr);
            }
        }
    }

    false
}

fn get_successor_map<'a>(
    container: &'a IrBlockContainer,
    successor_cache: &'a mut HashMap<u32, HashMap<u32, Vec<u32>>>,
) -> &'a HashMap<u32, Vec<u32>> {
    successor_cache
        .entry(container.id)
        .or_insert_with(|| build_successor_map(container))
}

fn build_successor_map(container: &IrBlockContainer) -> HashMap<u32, Vec<u32>> {
    let block_set = container
        .blocks
        .iter()
        .map(|x| x.start_address)
        .collect::<HashSet<_>>();

    let mut out = HashMap::<u32, Vec<u32>>::new();
    for (index, block) in container.blocks.iter().enumerate() {
        let next_block = container.blocks.get(index + 1).map(|x| x.start_address);
        let mut successors = Vec::<u32>::new();
        let Some(terminator) = get_block_terminator(block) else {
            if let Some(next_address) = next_block {
                successors.push(next_address);
            }
            out.insert(block.start_address, successors);
            continue;
        };

        for target in iter_block_targets(terminator) {
            if block_set.contains(&target) && !successors.contains(&target) {
                successors.push(target);
            }
        }
        if matches!(
            terminator,
            IrStatement::If(IrIf {
                false_statement: None,
                ..
            })
        ) && let Some(next_address) = next_block
            && !successors.contains(&next_address)
        {
            successors.push(next_address);
        }
        if matches!(
            terminator,
            IrStatement::Switch(IrSwitch {
                default_target: None,
                ..
            })
        ) && let Some(next_address) = next_block
            && !successors.contains(&next_address)
        {
            successors.push(next_address);
        }
        out.insert(block.start_address, successors);
    }

    out
}

fn writes_variable_top_level(statement: &IrStatement, address: u32) -> bool {
    matches!(statement, IrStatement::Assignment(IrAssignmentStatement { target, .. })
        if assignment_target_variable_address(target).is_some_and(|target_address| target_address == address))
}

fn count_reads_top_level(statement: &IrStatement, address: u32) -> usize {
    match statement {
        IrStatement::Assignment(IrAssignmentStatement { target, value }) => {
            count_assignment_target_reads(target, address) + count_expr_reads(value, address)
        }
        IrStatement::Expression(IrExpressionStatement { expression }) => {
            count_expr_reads(expression, address)
        }
        IrStatement::If(IrIf { condition, .. }) => count_expr_reads(condition, address),
        IrStatement::Switch(IrSwitch {
            index_expression, ..
        }) => count_expr_reads(index_expression, address),
        _ => 0,
    }
}

fn count_reads_in_statement(statement: &IrStatement, address: u32) -> usize {
    match statement {
        IrStatement::Assignment(IrAssignmentStatement { target, value }) => {
            count_assignment_target_reads(target, address) + count_expr_reads(value, address)
        }
        IrStatement::Expression(IrExpressionStatement { expression }) => {
            count_expr_reads(expression, address)
        }
        IrStatement::If(IrIf {
            condition,
            true_statement,
            false_statement,
        }) => {
            let mut count = count_expr_reads(condition, address);
            count += count_reads_in_statement(true_statement, address);
            if let Some(false_statement) = false_statement.as_ref() {
                count += count_reads_in_statement(false_statement, address);
            }
            count
        }
        IrStatement::Switch(IrSwitch {
            index_expression, ..
        }) => count_expr_reads(index_expression, address),
        IrStatement::Block(block) => block
            .statements
            .iter()
            .map(|x| count_reads_in_statement(x, address))
            .sum(),
        IrStatement::BlockContainer(container) => container
            .blocks
            .iter()
            .flat_map(|b| b.statements.iter())
            .map(|x| count_reads_in_statement(x, address))
            .sum(),
        IrStatement::HighLevelSwitch(IrHighLevelSwitch {
            index_expression,
            sections,
        }) => {
            let mut count = count_expr_reads(index_expression, address);
            count += sections
                .iter()
                .flat_map(|s| s.body.statements.iter())
                .map(|x| count_reads_in_statement(x, address))
                .sum::<usize>();
            count
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => body
            .blocks
            .iter()
            .flat_map(|b| b.statements.iter())
            .map(|x| count_reads_in_statement(x, address))
            .sum(),
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => body
            .blocks
            .iter()
            .flat_map(|b| b.statements.iter())
            .map(|x| count_reads_in_statement(x, address))
            .sum(),
        IrStatement::Jump(_) | IrStatement::Leave(_) | IrStatement::Return(IrReturn) => 0,
        _ => 0,
    }
}

fn count_writes_in_statement(statement: &IrStatement, address: u32) -> usize {
    match statement {
        IrStatement::Assignment(IrAssignmentStatement { target, .. }) => usize::from(
            assignment_target_variable_address(target)
                .is_some_and(|target_address| target_address == address),
        ),
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            let mut count = count_writes_in_statement(true_statement, address);
            if let Some(false_statement) = false_statement.as_ref() {
                count += count_writes_in_statement(false_statement, address);
            }
            count
        }
        IrStatement::Block(block) => block
            .statements
            .iter()
            .map(|x| count_writes_in_statement(x, address))
            .sum(),
        IrStatement::BlockContainer(container) => container
            .blocks
            .iter()
            .flat_map(|b| b.statements.iter())
            .map(|x| count_writes_in_statement(x, address))
            .sum(),
        IrStatement::HighLevelSwitch(switch_stmt) => switch_stmt
            .sections
            .iter()
            .flat_map(|s| s.body.statements.iter())
            .map(|x| count_writes_in_statement(x, address))
            .sum(),
        IrStatement::HighLevelWhile(while_stmt) => while_stmt
            .body
            .blocks
            .iter()
            .flat_map(|b| b.statements.iter())
            .map(|x| count_writes_in_statement(x, address))
            .sum(),
        IrStatement::HighLevelDoWhile(do_while_stmt) => do_while_stmt
            .body
            .blocks
            .iter()
            .flat_map(|b| b.statements.iter())
            .map(|x| count_writes_in_statement(x, address))
            .sum(),
        _ => 0,
    }
}

fn count_expr_reads(expression: &IrExpression, address: u32) -> usize {
    match expression {
        IrExpression::Variable(variable_expression) => {
            usize::from(variable_expression.address == address)
        }
        IrExpression::OperatorCall(IrOperatorCallExpression { arguments, .. }) => {
            arguments.iter().map(|x| count_expr_reads(x, address)).sum()
        }
        IrExpression::ExternalCall(IrExternalCallExpression { arguments, .. }) => {
            arguments.iter().map(|x| count_expr_reads(x, address)).sum()
        }
        IrExpression::PropertyAccess(IrPropertyAccessExpression { arguments, .. }) => {
            arguments.iter().map(|x| count_expr_reads(x, address)).sum()
        }
        IrExpression::ConstructorCall(IrConstructorCallExpression { arguments, .. }) => {
            arguments.iter().map(|x| count_expr_reads(x, address)).sum()
        }
        _ => 0,
    }
}

fn replace_reads_top_level(
    statement: &mut IrStatement,
    address: u32,
    replacement: &IrExpression,
) -> bool {
    match statement {
        IrStatement::Assignment(IrAssignmentStatement { target, value }) => {
            replace_assignment_target_reads(target, address, replacement)
                + replace_expr_reads(value, address, replacement)
                == 1
        }
        IrStatement::Expression(IrExpressionStatement { expression }) => {
            replace_expr_reads(expression, address, replacement) == 1
        }
        IrStatement::If(IrIf { condition, .. }) => {
            replace_expr_reads(condition, address, replacement) == 1
        }
        IrStatement::Switch(IrSwitch {
            index_expression, ..
        }) => replace_expr_reads(index_expression, address, replacement) == 1,
        _ => false,
    }
}

fn replace_expr_reads(
    expression: &mut IrExpression,
    address: u32,
    replacement: &IrExpression,
) -> usize {
    match expression {
        IrExpression::Variable(variable_expression) => {
            if variable_expression.address == address {
                *expression = replacement.clone();
                1
            } else {
                0
            }
        }
        IrExpression::OperatorCall(IrOperatorCallExpression { arguments, .. }) => arguments
            .iter_mut()
            .map(|x| replace_expr_reads(x, address, replacement))
            .sum(),
        IrExpression::ExternalCall(IrExternalCallExpression { arguments, .. }) => arguments
            .iter_mut()
            .map(|x| replace_expr_reads(x, address, replacement))
            .sum(),
        IrExpression::PropertyAccess(IrPropertyAccessExpression { arguments, .. }) => arguments
            .iter_mut()
            .map(|x| replace_expr_reads(x, address, replacement))
            .sum(),
        IrExpression::ConstructorCall(IrConstructorCallExpression { arguments, .. }) => arguments
            .iter_mut()
            .map(|x| replace_expr_reads(x, address, replacement))
            .sum(),
        _ => 0,
    }
}

fn assignment_target_variable_address(target: &IrExpression) -> Option<u32> {
    match target {
        IrExpression::Variable(variable) => Some(variable.address),
        _ => None,
    }
}

fn count_assignment_target_reads(target: &IrExpression, address: u32) -> usize {
    match target {
        IrExpression::Variable(_) => 0,
        IrExpression::PropertyAccess(_) => count_expr_reads(target, address),
        _ => 0,
    }
}

fn replace_assignment_target_reads(
    target: &mut IrExpression,
    address: u32,
    replacement: &IrExpression,
) -> usize {
    match target {
        IrExpression::Variable(_) => 0,
        IrExpression::PropertyAccess(_) => replace_expr_reads(target, address, replacement),
        _ => 0,
    }
}
